plugins/test/framework.cc

```diff
@@ -185,20 +195,30 @@ TestHttpContext::Result TestHttpContext::SendRequestHeaders(
 TestHttpContext::Result TestHttpContext::SendRequestBody(std::string body) {
   phase_logs_.clear();
   result_ = Result{};
-  body_buffer_.setOwned(std::move(body));
   current_callback_ = TestHttpContext::CallbackType::RequestBody;
+  // Don't invoke callbacks if immediate response has been sent.
+  if (immediate_response_) return std::move(result_);
```

Comment:
These short circuits should be one-liners IMO so they don't cause new conditionals below:

`if (immediate_response_) return Result{};`